gk/fib.c

```diff
@@ -60,43 +61,82 @@ gk_lpm_add_ipv6_route(uint8_t *ip,
  * callers must ensure that the entry is not being used.
  */
 static inline int
-initialize_ether_cache(struct ether_cache *eth_cache)
+clear_ether_cache(struct ether_cache *eth_cache)
 {
 	memset(eth_cache, 0, sizeof(*eth_cache));
+	seqlock_init(&eth_cache->lock);
 
 	return 0;
 }
 
-/*
- * Fill up the Ethernet cached header.
- * Note that the destination MAC address should be filled up by LLS.
- */
-static inline void
+static void
+gk_arp_and_nd_req_cb(const struct lls_map *map, void *arg,
+	__attribute__((unused))enum lls_reply_ty ty, int *pcall_again)
+{
+	struct ether_cache *eth_cache = arg;
+
+	/*
+	 * Deal with concurrency control by sequential lock
+	 * on the nexthop entry.
+	 */
+	write_seqlock(&eth_cache->lock);
+	ether_addr_copy(&map->ha, &eth_cache->eth_hdr.d_addr);
+	eth_cache->stale = map->stale;
+	write_sequnlock(&eth_cache->lock);
+
+	if (pcall_again != NULL)
+		*pcall_again = true;
+}
+
+/* Fill up the Ethernet cached header. */
+static inline int
 fill_up_ether_cache_locked(struct ether_cache *eth_cache,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+	struct ipaddr *addr, struct gatekeeper_if *iface, int lcore_id)
 {
+	write_seqlock(&eth_cache->lock);
 	eth_cache->stale = true;
 	rte_memcpy(&eth_cache->ip_addr, addr, sizeof(eth_cache->ip_addr));
 	eth_cache->eth_hdr.ether_type = addr->proto;
 	ether_addr_copy(&iface->eth_addr, &eth_cache->eth_hdr.s_addr);
 	eth_cache->ref_cnt = 1;
+	write_sequnlock(&eth_cache->lock);
+
+	if (addr->proto == ETHER_TYPE_IPv4) {
+		return hold_arp(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v4, lcore_id);
+	} else {
+		return hold_nd(gk_arp_and_nd_req_cb,
+			eth_cache, &addr->ip.v6, lcore_id);
+	}
 }
 
 static struct ether_cache *
 neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
-	struct ipaddr *addr, struct gatekeeper_if *iface)
+	struct ipaddr *addr, struct gatekeeper_if *iface, int lcore_id)
 {
 	int ret;
 	struct ether_cache *eth_cache = lookup_ether_cache(neigh, &addr->ip);
 	if (eth_cache != NULL) {
+		write_seqlock(&eth_cache->lock);
 		eth_cache->ref_cnt++;
+		write_sequnlock(&eth_cache->lock);
+
 		return eth_cache;
 	}
 
 	ret = rte_hash_add_key(neigh->hash_table, &addr->ip);
 	if (ret >= 0) {
+		int succ;
 		eth_cache = &neigh->cache_tbl[ret];
-		fill_up_ether_cache_locked(eth_cache, addr, iface);
+		succ = fill_up_ether_cache_locked(
```

Comment:
Notice that we added an entry calling `rte_hash_add_key()` before, but the entry is NOT initilized. So before `fill_up_ether_cache_locked()` returns, bad things can happen with readers being confused with an uninitialized entry. The solution is call `fill_up_ether_cache_locked()` before adding the key to the hash table; function `rte_hash_add_key_data()` seems to enable that.